### PR TITLE
feat: distributed execution of update statement

### LINKED ISSUE #13961: Feature: distributed execution of `update` statement
**Summary**

Execution of `update` statement can not utilize the resources of whole cluster yet, let fix this (like `delete` does)

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. Introduce distributed execution of update statement
2. Fix: deduplicate label is invalid in distributed mode 

Compared with `v1.2.250-nightly` in medium size warehouse:
```sh
#13971
bendsql --query='drop table update_target;'
bendsql --query='create table update_target(c1 int,c2 int);'
bendsql --query='insert into update_target select number,number from numbers(5000000);'
bendsql --query='insert into update_target select number,number from numbers(5000000);'
bendsql --query='ALTER warehouse update_distributed suspend;'
bendsql --query='use warehouse update_distributed;'
time bendsql --query='update update_target set c2 = 1 where c1 % 2 = 0;' # 0m5.577s
bendsql --query='select count(*) from update_target where (c1 % 2 = 0) and (c2 != 1);'

# v1.2.250-nightly
bendsql --query='drop table update_target;'
bendsql --query='create table update_target(c1 int,c2 int);'
bendsql --query='insert into update_target select number,number from numbers(5000000);'
bendsql --query='insert into update_target select number,number from numbers(5000000);'
bendsql --query='ALTER warehouse single suspend;'
bendsql --query='use warehouse single;'
time bendsql --query='update update_target set c2 = 1 where c1 % 2 = 0;' # 0m7.747s
bendsql --query='select count(*) from update_target where (c1 % 2 = 0) and (c2 != 1);'
```

- Closes #13961

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13971)

### CODE REVIEW

File: src/query/service/src/schedulers/fragments/plan_fragment.rs [hunk 1]
```diff
@@ -63,6 +64,7 @@ pub enum FragmentType {
     ReplaceInto,
     Compact,
     Recluster,
+    UpdateSource,
```
Comment (zhyass) [inline src/query/service/src/schedulers/fragments/plan_fragment.rs:67]: Rename to `Update` is better?

File: src/query/service/src/sessions/query_ctx.rs [hunk 2]
```diff
@@ -140,6 +140,19 @@ impl QueryContext {
         })
     }
 
+    pub fn create_from_query_level_shared(ctx: Arc<QueryContext>) -> Arc<QueryContext> {
```
Comment (BohuTANG) [inline src/query/service/src/sessions/query_ctx.rs:143]: Better to add comments for this function.

File: src/query/service/src/sessions/query_ctx.rs [hunk 3]
```diff
@@ -140,6 +140,19 @@ impl QueryContext {
...
+        Arc::new(QueryContext {
+            partition_queue: Arc::new(RwLock::new(VecDeque::new())),
+            version: format!("DatabendQuery {}", *DATABEND_COMMIT_VERSION),
+            mysql_version: format!("{}-{}", MYSQL_VERSION, *DATABEND_COMMIT_VERSION),
+            clickhouse_version: CLICKHOUSE_VERSION.to_string(),
+            shared: ctx.shared.clone(),
+            query_settings: ctx.query_settings.clone(),
+            fragment_id: Arc::new(AtomicUsize::new(0)),
+            inserted_segment_locs: Arc::new(RwLock::new(HashSet::new())),
```
Comment (zhang2014) [inline src/query/service/src/sessions/query_ctx.rs:152]: move to context shared is better?

**Comment by SkyFan2002** [comment]
After discussing with @zhang2014, it'd be better that #13981 be solved in another PR, temporarily convert this PR to draft.

**Comment by SkyFan2002** [comment]
Get deduplicate_label early to make it work in distributed mode, which skips #13981 for now. Let's merge this PR cc @zhang2014 @dantengsky

**Review by dantengsky — APPROVED** [review]
LGTM

**Comment by BohuTANG** [comment]
Can we add a explain of the distributed update in another PR?

**Comment by SkyFan2002** [comment]
> Can we add a explain of the distributed update in another PR?

Of course

**Pull request merged**